### PR TITLE
Optimize Label parsing calls

### PR DESCRIPTION
**Pull request by Vlatombe**
Optimize Label parsing related calls

* `Node#getAssignedLabels` parse Labels from scratch on every call. This ends up being called a lot via `Jenkins#trimLabels`. Caching the set of `LabelAtom` allows to parse labels only if the label string is updated.
* `Label#parse` has come up in performance testing involving a lot of cloud agents using one-shot labels.

This speeds up lookup of an existing simple label (no expression used):  `16 ns` vs. `106 ns` (6x faster)

`Node#getAssignedLabels`
* `LabelBenchmark.jenkinsGetAssignedLabels` :  `34ns` vs. `41ns`, expected since Jenkins doesn't have any assigned label by default.
* `NodeLabelBenchmark.nodeGetAssignedLabels` : `91ns` vs. `185ns` (2x faster)

### Proposed changelog entries

- Performance optimizations on Label parsing.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8395"></a>

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java:2121]: `NoExternalUse`?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java:2122]: Existing caller will not in fact pass null.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java:2123]: Will this not throw a CCE if it is some other kind of `Label`?

AFAICT https://github.com/jenkinsci/jenkins/blob/a17f0f63c9fac78d3afc4f7864aa0399d6f1c7b4/core/src/main/java/jenkins/model/Jenkins.java#L2100-L2106 is also unsafe in general. I must be missing something.

```diff
diff --git test/src/test/java/hudson/model/labels/LabelExpressionTest.java test/src/test/java/hudson/model/labels/LabelExpressionTest.java
index 4716daa2ba..049a0a7f96 100644
--- test/src/test/java/hudson/model/labels/LabelExpressionTest.java
+++ test/src/test/java/hudson/model/labels/LabelExpressionTest.java
@@ -241,6 +241,12 @@ public class LabelExpressionTest {
         assertEquals("label1||label2", l.getName());
     }
 
+    @Test
+    public void wtf() {
+        j.jenkins.getLabel("xxx && yyy");
+        j.jenkins.getLabelAtom("xxx && yyy");
+    }
+
     /**
      * The name should have parenthesis at the right place to preserve the tree structure.
      */
```

⇒

```
java.lang.ClassCastException: class hudson.model.labels.LabelExpression$And cannot be cast to class hudson.model.labels.LabelAtom (hudson.model.labels.LabelExpression$And and hudson.model.labels.LabelAtom are in unnamed module of loader 'app')
	at jenkins.model.Jenkins.getLabelAtom(Jenkins.java:2106)
	at hudson.model.labels.LabelExpressionTest.wtf(LabelExpressionTest.java:247)
```

I guess `getLabelAtom` (and `LabelAtom.get`) is just called from contexts where an atom is actually expected; whereas this new method is called from `Label.parse` and would routinely be given a label expression?

File: test/src/test/java/hudson/model/labels/LabelBenchmark.java
Comment (jglick) [inline test/src/test/java/hudson/model/labels/LabelBenchmark.java:10]: Can this be a nested class, to avoid proliferation of source files?

**Review by NotMyFault — COMMENTED** [review]
Spotbugs and the failure in `hudson.XmlFileTest.canReadXmlWithControlCharsTest` are left to address

**Review by NotMyFault — APPROVED** [review]
Thanks!

**Review by jglick — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/hudson/model/Node.java
Comment (jglick) [inline core/src/main/java/hudson/model/Node.java:305]: This could also be restricted I suppose. (In practice I do not know of any `Node` subclasses besides `Jenkins` and `Slave`.)

**Review by jglick — APPROVED** [review]

**Comment by MarkEWaite** [comment]
@Vlatombe thanks for this improvement.  It appears that the improvement causes one of the automated tests in the workflow durable task plugin to fail as shown in https://github.com/jenkinsci/bom/pull/2412/checks?check_run_id=16175702436 

Thanks to @daniel-beck for detecting the issue and mentioning it in https://github.com/jenkinsci/bom/pull/2412#issuecomment-1691512328

Can you investigate that failure and propose a fix for the tests in the workflow durable task plugin so that a new version of workflow durable task plugin can be included in the plugin bill of materials?  Without that fix, the plugin compatibility tests used for the plugin bill of materials will fail and block the next release of the plugin bill of materials.

**Comment by Vlatombe** [comment]
Hi @MarkEWaite this test failure has been addressed already - https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/335

**Comment by MarkEWaite** [comment]
Thanks very much @Vlatombe !  I've triggered a dependabot update of the plugin bill of materials to bring the new release of the plugin into the BOM.

**Comment by aaronmell** [comment]
FYI, this change likely broken our Jenkins instance.  Jenkins fails to start after upgrading, and we are getting a null reference error

```
Error
java.lang.NullPointerException
	at java.base/java.util.HashSet.<init>(HashSet.java:119)
	at hudson.model.Node.getAssignedLabels(Node.java:326)
	at jenkins.model.Jenkins.lambda$trimLabels$3(Jenkins.java:2323)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at jenkins.model.Jenkins.trimLabels(Jenkins.java:2323)
	at jenkins.model.Jenkins.trimLabels(Jenkins.java:2303)
	at jenkins.model.Nodes$6.run(Nodes.java:352)
	at hudson.model.Queue._withLock(Queue.java:1397)
	at hudson.model.Queue.withLock(Queue.java:1271)
	at jenkins.model.Nodes.load(Nodes.java:346)
	at jenkins.model.Jenkins$12.run(Jenkins.java:3488)
	at org.jvnet.hudson.reactor.TaskGraphBuilder$TaskImpl.run(TaskGraphBuilder.java:177)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:305)
	at jenkins.model.Jenkins$5.runTask(Jenkins.java:1170)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:221)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:120)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused: org.jvnet.hudson.reactor.ReactorException
	at org.jvnet.hudson.reactor.Reactor.execute(Reactor.java:290)
	at jenkins.InitReactorRunner.run(InitReactorRunner.java:49)
	at jenkins.model.Jenkins.executeReactor(Jenkins.java:1205)
	at jenkins.model.Jenkins.<init>(Jenkins.java:992)
	at hudson.model.Hudson.<init>(Hudson.java:86)
	at hudson.model.Hudson.<init>(Hudson.java:82)
	at hudson.WebAppMain$3.run(WebAppMain.java:247)
Caused: hudson.util.HudsonFailedToLoad
	at hudson.WebAppMain$3.run(WebAppMain.java:264)
```

**Comment by aaronmell** [comment]
We fixed our issue by removing all of the nodes in the nodes folder

**Comment by Vlatombe** [comment]
Filed https://github.com/jenkinsci/ec2-plugin/pull/883 to fix the impacted plugin. Given this happened already with 2 plugins overriding `readResolve` without calling its super implementation, I will also file a robustness improvement for this change.

**Comment by Vlatombe** [comment]
Filed https://github.com/jenkinsci/jenkins/pull/8448

File: core/src/main/java/hudson/model/Slave.java
Comment (Vlatombe) [inline core/src/main/java/hudson/model/Slave.java:199]: Caused [JENKINS-71937](https://issues.jenkins.io/browse/JENKINS-71937), as `label` doesn't get set

**Pull request merged**
